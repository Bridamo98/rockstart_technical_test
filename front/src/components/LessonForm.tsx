import { zodResolver } from "@hookform/resolvers/zod";
import { useForm } from "react-hook-form";
import * as z from "zod";

import { extractYoutubeId } from "../utils/youtube";

const isYoutubeUrl = (url: string) => !!extractYoutubeId(url);

const createSchema = z.object({
  title: z
    .string()
    .min(1, "Title is required")
    .max(255, "Title must be â‰¤ 255 characters"),
  video_url: z
    .string()
    .url("Invalid URL")
    .refine(isYoutubeUrl, { message: "Must be a YouTube link" }),
});
const updateSchema = createSchema.partial();

type CreateFormData = z.infer<typeof createSchema>;
type UpdateFormData = z.infer<typeof updateSchema>;
type FormData = CreateFormData | UpdateFormData;

interface Props {
  defaultValues?: UpdateFormData;
  onSubmit: (values: FormData) => void | Promise<void>;
  submitLabel?: string;
}

export default function LessonForm({
  defaultValues,
  onSubmit,
  submitLabel = "Save",
}: Props) {
  const isEdit = !!defaultValues;
  const schema = isEdit ? updateSchema : createSchema;

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormData>({
    defaultValues,
    resolver: zodResolver(schema),
  });

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
      className="space-y-4 bg-gray-100 p-4 rounded"
    >
      <div>
        <input
          {...register("title")}
          maxLength={255}
          placeholder="Lesson title"
          className="input w-full"
        />
        {errors.title && (
          <p className="text-sm text-red-600">{errors.title.message}</p>
        )}
      </div>

      <div>
        <input
          {...register("video_url")}
          placeholder="YouTube URL"
          className="input w-full"
        />
        {errors.video_url && (
          <p className="text-sm text-red-600">{errors.video_url.message}</p>
        )}
      </div>

      <button className="btn btn-primary">{submitLabel}</button>
    </form>
  );
}
