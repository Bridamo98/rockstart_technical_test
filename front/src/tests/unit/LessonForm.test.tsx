import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { expect, it, vi } from "vitest";
import LessonForm from "../../components/LessonForm";

it("rechaza url que no sea de YouTube (snapshot)", async () => {
  const { asFragment } = render(<LessonForm onSubmit={vi.fn()} />);
  const user = userEvent.setup();

  await user.type(screen.getByPlaceholderText(/lesson title/i), "L1");
  await user.type(
    screen.getByPlaceholderText(/youtube url/i),
    "https://vimeo.com/123"
  );
  await user.click(screen.getByRole("button", { name: /save/i }));

  expect(
    await screen.findByText(/must be a youtube link/i)
  ).toBeInTheDocument();

  expect(asFragment()).toMatchSnapshot();
});
