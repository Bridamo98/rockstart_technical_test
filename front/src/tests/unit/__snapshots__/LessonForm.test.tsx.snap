// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rechaza url que no sea de YouTube (snapshot) 1`] = `
<DocumentFragment>
  <form
    class="space-y-4 bg-gray-100 p-4 rounded"
  >
    <div>
      <input
        class="input w-full"
        maxlength="255"
        name="title"
        placeholder="Lesson title"
      />
    </div>
    <div>
      <input
        class="input w-full"
        name="video_url"
        placeholder="YouTube URL"
      />
      <p
        class="text-sm text-red-600"
      >
        Must be a YouTube link
      </p>
    </div>
    <button
      class="btn btn-primary"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;
